From 657720333ce23b3ef639bd6d1df24ee6b6b51768 Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <bmastbergen@ciq.com>
Date: Tue, 9 Sep 2025 10:21:26 -0400
Subject: [PATCH] drivers:md:fix a potential use-after-free bug

jira VULN-135015
cve CVE-2022-50022
commit-author Wentao_Liang <Wentao_Liang_g@163.com>
commit 104212471b1c1817b311771d817fb692af983173

In line 2884, "raid5_release_stripe(sh);" drops the reference to sh and
may cause sh to be released. However, sh is subsequently used in lines
2886 "if (sh->batch_head && sh != sh->batch_head)". This may result in an
use-after-free bug.

It can be fixed by moving "raid5_release_stripe(sh);" to the bottom of
the function.

	Signed-off-by: Wentao_Liang <Wentao_Liang_g@163.com>
	Signed-off-by: Song Liu <song@kernel.org>
	Signed-off-by: Jens Axboe <axboe@kernel.dk>
(cherry picked from commit 104212471b1c1817b311771d817fb692af983173)
	Signed-off-by: Brett Mastbergen <bmastbergen@ciq.com>
---
 drivers/md/raid5.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/md/raid5.c b/drivers/md/raid5.c
index 730558b79fa1..c0f9a15de2d6 100644
--- a/drivers/md/raid5.c
+++ b/drivers/md/raid5.c
@@ -2678,10 +2678,10 @@ static void raid5_end_write_request(struct bio *bi, int error)
 	if (!test_and_clear_bit(R5_DOUBLE_LOCKED, &sh->dev[i].flags))
 		clear_bit(R5_LOCKED, &sh->dev[i].flags);
 	set_bit(STRIPE_HANDLE, &sh->state);
-	raid5_release_stripe(sh);
 
 	if (sh->batch_head && sh != sh->batch_head)
 		raid5_release_stripe(sh->batch_head);
+	raid5_release_stripe(sh);
 }
 
 static void raid5_error(struct mddev *mddev, struct md_rdev *rdev)
-- 
2.47.3

